File: client/android/app/src/main/java/com/stonefive/chalkak/feature/signature/SignaturePngEncoder.kt
package com.stonefive.chalkak.feature.signature

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import java.io.ByteArrayOutputStream
import kotlin.math.min

fun interface SignaturePngEncoder {
    fun encode(strokes: List<SignatureStroke>): ByteArray
}

class AndroidSignaturePngEncoder : SignaturePngEncoder {
    override fun encode(strokes: List<SignatureStroke>): ByteArray {
        val drawableStrokes = strokes.filter { it.points.isNotEmpty() }
        require(drawableStrokes.isNotEmpty()) { "저장할 사인이 없습니다." }

        val points = drawableStrokes.flatMap(SignatureStroke::points)
        val minX = points.minOf(SignaturePoint::xRatio)
        val maxX = points.maxOf(SignaturePoint::xRatio)
        val minY = points.minOf(SignaturePoint::yRatio)
        val maxY = points.maxOf(SignaturePoint::yRatio)
        val contentWidth = (maxX - minX).coerceAtLeast(MIN_NORMALIZED_SIZE)
        val contentHeight = (maxY - minY).coerceAtLeast(MIN_NORMALIZED_SIZE)
        val availableWidth = OUTPUT_WIDTH - PADDING_PX * 2f
        val availableHeight = OUTPUT_HEIGHT - PADDING_PX * 2f
        val scale = min(availableWidth / contentWidth, availableHeight / contentHeight)
        val offsetX = (OUTPUT_WIDTH - contentWidth * scale) / 2f
        val offsetY = (OUTPUT_HEIGHT - contentHeight * scale) / 2f

        val bitmap = Bitmap.createBitmap(OUTPUT_WIDTH, OUTPUT_HEIGHT, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        canvas.drawColor(Color.TRANSPARENT)
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = SIGNATURE_COLOR
            style = Paint.Style.STROKE
            strokeWidth = STROKE_WIDTH_PX
            strokeCap = Paint.Cap.ROUND
            strokeJoin = Paint.Join.ROUND
        }

        drawableStrokes.forEach { stroke ->
            val mappedPoints = stroke.points.map { point ->
                MappedPoint(
                    x = offsetX + (point.xRatio - minX) * scale,
                    y = offsetY + (point.yRatio - minY) * scale,
                )
            }
            if (mappedPoints.size == 1) {
                val point = mappedPoints.first()
                canvas.drawCircle(
                    point.x,
                    point.y,
                    STROKE_WIDTH_PX / 2f,
                    paint.apply {
                        style = Paint.Style.FILL
                    },
                )
                paint.style = Paint.Style.STROKE
            } else {
                canvas.drawPath(mappedPoints.toSmoothPath(), paint)
            }
        }

        return try {
            ByteArrayOutputStream().use { output ->
                check(bitmap.compress(Bitmap.CompressFormat.PNG, 100, output)) {
                    "사인 이미지를 PNG로 변환하지 못했습니다."
                }
                output.toByteArray()
            }
        } finally {
            bitmap.recycle()
        }
    }

    private fun List<MappedPoint>.toSmoothPath(): Path = Path().apply {
        moveTo(first().x, first().y)
        for (index in 1 until lastIndex) {
            val current = this@toSmoothPath[index]
            val next = this@toSmoothPath[index + 1]
            quadTo(
                current.x,
                current.y,
                (current.x + next.x) / 2f,
                (current.y + next.y) / 2f,
            )
        }
        lineTo(last().x, last().y)
    }

    private data class MappedPoint(
        val x: Float,
        val y: Float,
    )

    private companion object {
        const val OUTPUT_WIDTH = 1024
        const val OUTPUT_HEIGHT = 512
        const val PADDING_PX = 48
        const val STROKE_WIDTH_PX = 14f
        const val MIN_NORMALIZED_SIZE = 0.02f
        val SIGNATURE_COLOR = Color.WHITE
    }
}

File: client/iOS/chalkak/chalkak/Features/Signature/SignaturePngEncoder.swift
import CoreGraphics
import UIKit

struct SignaturePngEncoder {
    func encode(_ strokes: [SignatureStroke]) throws -> Data {
        let drawableStrokes = strokes.filter { !$0.isEmpty }
        guard !drawableStrokes.isEmpty else {
            throw SignaturePngEncodingError.emptySignature
        }

        let points = drawableStrokes.flatMap(\.points)
        let minX = points.map(\.xRatio).min() ?? 0
        let maxX = points.map(\.xRatio).max() ?? 0
        let minY = points.map(\.yRatio).min() ?? 0
        let maxY = points.map(\.yRatio).max() ?? 0
        let contentWidth = max(maxX - minX, Metrics.minimumNormalizedSize)
        let contentHeight = max(maxY - minY, Metrics.minimumNormalizedSize)
        let availableWidth = Metrics.outputWidth - Metrics.padding * 2
        let availableHeight = Metrics.outputHeight - Metrics.padding * 2
        let scale = min(availableWidth / contentWidth, availableHeight / contentHeight)
        let offsetX = (Metrics.outputWidth - contentWidth * scale) / 2
        let offsetY = (Metrics.outputHeight - contentHeight * scale) / 2

        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        format.opaque = false
        let renderer = UIGraphicsImageRenderer(
            size: CGSize(width: Metrics.outputWidth, height: Metrics.outputHeight),
            format: format
        )

        return renderer.pngData { rendererContext in
            let context = rendererContext.cgContext
            context.setStrokeColor(UIColor.white.cgColor)
            context.setFillColor(UIColor.white.cgColor)
            context.setLineWidth(Metrics.strokeWidth)
            context.setLineCap(.round)
            context.setLineJoin(.round)

            for stroke in drawableStrokes {
                let mappedPoints = stroke.points.map { point in
                    CGPoint(
                        x: offsetX + (point.xRatio - minX) * scale,
                        y: offsetY + (point.yRatio - minY) * scale
                    )
                }
                guard let firstPoint = mappedPoints.first else { continue }
                if mappedPoints.count == 1 {
                    let radius = Metrics.strokeWidth / 2
                    context.fillEllipse(
                        in: CGRect(
                            x: firstPoint.x - radius,
                            y: firstPoint.y - radius,
                            width: radius * 2,
                            height: radius * 2
                        )
                    )
                } else {
                    context.addPath(mappedPoints.smoothPath())
                    context.strokePath()
                }
            }
        }
    }
}

enum SignaturePngEncodingError: Error {
    case emptySignature
}

private extension Array where Element == CGPoint {
    func smoothPath() -> CGPath {
        let path = CGMutablePath()
        path.move(to: self[0])
        if count == 2 {
            path.addLine(to: self[1])
            return path
        }
        for index in 1..<(count - 1) {
            let current = self[index]
            let next = self[index + 1]
            path.addQuadCurve(
                to: CGPoint(x: (current.x + next.x) / 2, y: (current.y + next.y) / 2),
                control: current
            )
        }
        path.addLine(to: self[count - 1])
        return path
    }
}

private enum Metrics {
    static let outputWidth: CGFloat = 1024
    static let outputHeight: CGFloat = 512
    static let padding: CGFloat = 48
    static let strokeWidth: CGFloat = 14
    static let minimumNormalizedSize: CGFloat = 0.02
}

